pytorch_lightning/trainer/trainer.py

```diff
@@ -1164,6 +1164,12 @@ def _run_train(self) -> None:
 
         self.reset_train_val_dataloaders(model)
 
+        # fresh start epoch can be under following situations:
```

Comment:
Hey @ninginthecloud.

Quick idea to investigate. What if we remove the `reset_train_val_dataloaders` from the Trainer #1165 and left only the loop handle the reloading logic. This is a blocker for loop customization.

Best,
T.C

Reply:
thanks for the suggestion, @tchaton ! That's  what I've mentioned in the description as option 2 for resolving this fix. We definitely can go with that route,and it will change a lot of existing underlying assumptions and behaviors. 

> **Option 2**: remove `self.reset_train_val_dataloaders(model)` in `Trainer._run_train()` and set the reload logics in corresponding loops. When I worked on this option, I found a lot of other underlying assumptions we need to fix:
> 1. `self.num_training_batches = 0` in `trainer._setup_on_init()`. This value should be initialized as `self.num_training_batches = float('inf')`, otherwise, if we remove dataloader reloading in `Trainer._run_train()` before calling fit_loop.run(), the `skip` flag from `fit_loop` will be true. which makes `fit_loop` falsely stops. 
> 2. we allow users to call `trainer.fit()` multiple times with different dataloaders. removing dataloader reloading in `Trainer._run_train()`, it marks `fit_loop` unable to continue fitting with new data. In this sense, we still need  attribute `_is_fresh_start_epoch` mentioned in Option 1. 
> 3. `reset_val_dataloader()` is called before fit_loop officially starts. Sometimes it's called if sanity_check is enabled. naively removing `self.reset_train_val_dataloaders(model)` in `Trainer._run_train()` breaks a lot of tests since our usage and assumption of loading `val_dataloder()` changed. As mentioned in issue #8738, it's better to provide more gradually control of reloading train and val dataloader seperately. 
> In summary, this PR adopts the option 1 to quickly fix this bug. In the next step, let's complete issue #8738 and then refactor dataloader reloading.

We agree that ultimately we want to provide loop customization, after all, it does not make sense to let trainer control  data reloading. Let me know what you think. I can either 1) postpone issue #8738 to complete option 2, or adopt option 1 and complete #8738 and then refactor.

Reply:
IMO if #9502 isn't blocking something important, we can go the longer route for option 2 and wait this out.

Reply:
great~ Thanks @carmocca and @tchaton~ Let me update this PR soon.

Reply:
Thanks @ninginthecloud